espnet2/asr/encoder/contextual_block_transformer_encoder.py

```diff
@@ -325,3 +347,342 @@ def forward(
+
+            prev_addin = addin
+
+            
+        # set up masks
+        mask_online = xs_pad.new_zeros(
+            xs_pad.size(0), block_num, self.block_size + 2, self.block_size + 2
+        )
+        mask_online.narrow(2, 1, self.block_size + 1).narrow(
+            3, 0, self.block_size + 1
+        ).fill_(1)
+
+        # forward
+        if False:
+            ys_chunks=[]
+            for i in range(block_num):
+                print("loop ", i)
+                cur_mask_online = mask_online.narrow(1, i, 1)
+                cur_xs_chunk = xs_chunk.narrow(1, i, 1)
+                ys_chunk, _, _, past_encoder_ctx, _,_ = self.encoders(cur_xs_chunk, cur_mask_online, past_encoder_ctx)
+                ys_chunks.append(ys_chunk)
+                
+            ys_chunk = torch.cat(ys_chunks, dim=1)
+        else:        
+            ys_chunk, mask_online, _, past_encoder_ctx, _,_ = self.encoders(xs_chunk, mask_online, past_encoder_ctx)
+        #print(xs_chunk.shape, mask_online.shape, ys_chunk.shape)
+        #import sys
+        #sys.exit(1)
+        # remove addin
+        ys_chunk = ys_chunk.narrow(2, 1, self.block_size)
+
+        offset = self.block_size - self.look_ahead - self.hop_size
+        if is_final:
+            if n_processed_blocks == 0:
+                y_length = xs_pad.size(1)
+            else:
+                y_length = xs_pad.size(1) - offset
+        else:
+            y_length = block_num * self.hop_size
+            if n_processed_blocks == 0:
+                y_length += offset
+        ys_pad = xs_pad.new_zeros((xs_pad.size(0), y_length, xs_pad.size(2)))
+        if n_processed_blocks == 0:
+            ys_pad[:,0:offset] = ys_chunk[:,0,0:offset]
+        for i in range(block_num):
+            cur_hop=i*self.hop_size
+            if n_processed_blocks == 0:
+                cur_hop += offset
+            if i == block_num-1 and is_final:
+                chunk_length = min(self.block_size-offset, ys_pad.size(1)-cur_hop)
+            else:
+                chunk_length = self.hop_size
+            ys_pad[:, cur_hop:cur_hop+chunk_length] = ys_chunk[
+                :, i, offset:offset+chunk_length
+            ]
+        if self.normalize_before:
+            ys_pad = self.after_norm(ys_pad)
+
+        olens = masks.squeeze(1).sum(1)
+
+        if is_final:
+            next_states = None
+        else:
+            next_states = {
+                "prev_addin":prev_addin,
+                "buffer_before_downsampling": buffer_before_downsampling,
+                "ilens_buffer": ilens_buffer,
+                "buffer_after_downsampling": buffer_after_downsampling,
+                "masks_buffer": masks_buffer,
+                "n_processed_blocks": n_processed_blocks+block_num,
+                "past_encoder_ctx": past_encoder_ctx
+            }
+        
+        return ys_pad, olens, next_states
+
+    # Refactored implementation of Emiru's code
+    def forward_train2(
```

Comment:
@hfujihara, Is this used?
Do you mean that it is better to use this?

Reply:
This is not used and I think we could delete this for this PR.
Maybe in the future, it would be better if we could replace this with forward_train (Emiru's original implementation), but to do this it needs to be tested more to make sure it works exactly the same way in training.

Reply:
Yes, I have deleted this in the resubmitted PR. I'm sorry I make some mistakes when submitting the first PR.